### PR TITLE
:sparkles: Refactor wrappers & PP, Add Checkpoint Ensembles, EMA, SWA, & SWAG, Add LaplaceApprox & ABNN

### PR DESCRIPTION
**Pull request by o-laurent**
- [x] Add Helmholtz tutorials
- [x] Add ABNN
- [x] Merge #101 
- [x] Merge #96 
- [x] Enable choosing the post-processing method

### CODE REVIEW

File: auto_tutorials_source/tutorial_bayesian.py [hunk 1]
```diff
@@ -125,8 +126,10 @@ def optim_lenet(model: nn.Module) -> dict:
 # 6. Testing the Model
 # ~~~~~~~~~~~~~~~~~~~~
 #
-# Now that the model is trained, let's test it on MNIST
-
+# Now that the model is trained, let's test it on MNIST.
+# Please note that we apply a reshape to the logits to determine the dimension corresponding to the ensemble
+# and to the batch. As for TorchUncertainty 2.0, the ensemble dimension is merged with the batch dimension
```
Comment (o-laurent) [inline auto_tutorials_source/tutorial_bayesian.py:131]: 0.2.0

File: experiments/classification/mnist/configs/lenet_checkpoint_ensemble.yaml [hunk 2]
```diff
@@ -0,0 +1,73 @@
+# lightning.pytorch==2.1.3
```
Comment (o-laurent) [inline experiments/classification/mnist/configs/lenet_checkpoint_ensemble.yaml:1]: Add the config to the main ReadMe

File: tests/models/wrappers/test_stochastic.py [hunk 3]
```diff
@@ -47,27 +45,31 @@ class TestStochasticModel:
...
-        assert not model.layer.frozen
+        assert not model.model.layer.frozen
+        model.eval()
+        model(torch.randn(1, 1))
 
-        model = DummyModelConv()
+        model = StochasticModel(DummyModelConv(), 2)
         model.freeze()
-        assert model.layer.frozen
+        assert model.model.layer.frozen
```
Comment (alafage) [inline tests/models/wrappers/test_stochastic.py:59]: Maybe try renaming "model" as "inner_model" or "core_model"

File: torch_uncertainty/baselines/classification/vgg.py [hunk 4]
```diff
@@ -175,15 +162,9 @@ def __init__(
             num_classes=num_classes,
             model=model,
             loss=loss,
-            num_estimators=num_estimators,
+            is_ensemble=version in self.ensemble,
```
Comment (alafage) [inline torch_uncertainty/baselines/classification/vgg.py:165]: Rename self.ensemble into something clearer

File: torch_uncertainty/datamodules/abstract.py [hunk 5]
```diff
@@ -148,13 +150,13 @@ def make_cross_val_splits(
         return cv_dm
 
 
-class CrossValDataModule(AbstractDataModule):
+class CrossValDataModule(BaseDataModule):
```
Comment (alafage) [inline torch_uncertainty/datamodules/abstract.py:153]: Rename all Base[...] objects with TU[...]

File: torch_uncertainty/datasets/classification/not_mnist.py [hunk 6]
```diff
@@ -66,7 +66,7 @@ def __init__(
             )
 
         super().__init__(
-            self.root + f"/notMNIST_{subset}",
+            self.root / ("notMNIST_" + subset),
```
Comment (alafage) [inline torch_uncertainty/datasets/classification/not_mnist.py:69]: Use format string instead?

File: torch_uncertainty/layers/bayesian/sampler.py [hunk 7]
```diff
@@ -43,26 +41,27 @@ def log_posterior(self, weight: Tensor | None = None) -> Tensor:
         return -lposterior.sum()
 
 
-class PriorDistribution(nn.Module):
+class CenteredGaussianMixture(nn.Module):
```
Comment (alafage) [inline torch_uncertainty/layers/bayesian/sampler.py:44]: Should inherit from torch.distributions.Distribution?

File: torch_uncertainty/layers/distributions.py [hunk 8]
```diff
@@ -1,22 +1,25 @@
...
+
 import torch.nn.functional as F
 from torch import Tensor, nn
 from torch.distributions import Distribution, Laplace, Normal
 
 from torch_uncertainty.utils.distributions import NormalInverseGamma
 
 
-class _AbstractDist(nn.Module):
+class AbstractDist(ABC, nn.Module):
```
Comment (alafage) [inline torch_uncertainty/layers/distributions.py:10]: Rename to TUDist?

**Review by alafage — CHANGES_REQUESTED** [review]
See comments

File: auto_tutorials_source/tutorial_der_cubic.py [hunk 9]
```diff
@@ -55,9 +55,7 @@ def optim_regression(
         lr=learning_rate,
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply auto_tutorials_source/tutorial_der_cubic.py:55]: In the end, we removed it as it makes the code grow heavy in tutorials.

File: torch_uncertainty/routines/segmentation.py [hunk 10]
```diff
@@ -210,16 +231,10 @@ def subsample(self, pred: Tensor, target: Tensor) -> tuple[Tensor, Tensor]:
...
     num_classes: int,
     metric_subsampling_rate: float,
     num_calibration_bins: int,
 ) -> None:
-    if num_estimators < 1:
-        raise ValueError(
-            f"num_estimators must be positive, got {num_estimators}."
-        )
-
     if num_classes < 2:
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply torch_uncertainty/routines/segmentation.py:238]: moved to next PR

File: auto_tutorials_source/tutorial_bayesian.py [hunk 11]
```diff
@@ -55,12 +55,12 @@
 # We will use the Adam optimizer with the default learning rate of 0.001.
 
 
-def optim_lenet(model: nn.Module) -> dict:
+def optim_lenet(model: nn.Module):
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply auto_tutorials_source/tutorial_bayesian.py:58]: Type removed

File: torch_uncertainty/layers/bayesian/sampler.py (hunk 7 above)
Comment (o-laurent) [reply torch_uncertainty/layers/bayesian/sampler.py:44]: It's not possible for now, but we'll try in the following PR...

**Review by alafage — APPROVED** [review]

**Pull request merged**